Introduce batch action feature

yceruto opened the pull request:
Implementing https://github.com/EasyCorp/EasyAdminBundle/projects/1#card-846124

This is how it looks:
![ea-batch-actions](https://user-images.githubusercontent.com/2028198/51131918-f7eac100-17fe-11e9-86b8-22a78d88af0e.gif)

## Config

By default, batch actions are disabled (no default).

Configure a batch action very similar to normal action by using `batch_actions` option:
```yaml
easy_admin:
    list:
        batch_actions: ['delete']
    
    entities:
        Product:
            class: App\Entity\Product
            controller: App\Controller\ProductController
            list:
                batch_actions: ['restock'] # [{name: 'restock', label: 'Restock', icon: 'plus-square'}]
```
Now you can define the ``restockBatchAction()`` method in your own controller (the method will receive the selected items through `array $ids` argument):
```php
class ProductController extends EasyAdminController
{
    protected function restockBatchAction(array $ids): void
    {
        $this->getDoctrine()->getManagerForClass(Product::class)
            ->createQueryBuilder()
            ->update()
            ->from(Product::class, 'p')
            ->set('p.stock', 'p.stock + 10')
            ->where('p.id IN (:ids)')
            ->setParameter('ids', $ids)
            ->getQuery()
            ->execute()
        ;
    }
}
```
Also, as is common in EasyAdmin, you can define it by prefixing the entity name `restock<EntityName>BatchAction()`.

Done!

Try this feature (demo project) https://github.com/yceruto/ea2_batch_actions, see https://github.com/EasyCorp/EasyAdminBundle/pull/2578#issuecomment-455564697 for custom batch form.

## Notes
* For `delete` batch action you'll have a built-in implementation, deleting all selected items by default.
* Batch actions only support `method` type "by now". e.g. setting `{ name: 'restock', 'type': 'route' }` will throw an exception.
* Batch actions don't support composite primary key, it uses always the single identifier for selection.

## TODO

If the proposal/behavior/code looks good to you I'll add these missing topics:
- [x] CSS tweaks for `.batch_actions` class
- [ ] Tests
- [ ] Docs

Cheers!

alterphp commented on src/Controller/AdminControllerTrait.php:
```diff
@@ -387,6 +387,36 @@ protected function searchAction()
+    {
+        $request = $this->request;
+
+        if ($request->isMethod('BATCH') && $this->isCsrfTokenValid('batch_action', $request->request->get('_token'))) {
```
Why not a POST request here ? This is standard and, IMO, the best matching method for the command being sent through the request.

By the way, this implies `framework.http_method_override` Symfony config to be set to true (https://symfony.com/doc/current/reference/configuration/framework.html#configuration-framework-http-method-override).

yceruto replied:
No strong opinion about this. I like that the request method matches the request action, but you're right about standards, so fixed.

alterphp commented on src/Controller/AdminControllerTrait.php:
```diff
@@ -387,6 +387,36 @@ protected function searchAction()
+        if ($request->isMethod('BATCH') && $this->isCsrfTokenValid('batch_action', $request->request->get('_token'))) {
+            $batchActionName = $request->request->get('batch_action_name');
+
+            if ($batchActionName && $ids = explode('|', $request->request->get('batch_action_ids'))) {
```
I would use `,` as it is a more natural array separator to me but this is a pure opinionated feeling.

alterphp commented on src/Controller/AdminControllerTrait.php:
```diff
@@ -387,6 +387,36 @@ protected function searchAction()
+
+            if ($batchActionName && $ids = explode('|', $request->request->get('batch_action_ids'))) {
+                $class = $this->entity['class'];
+                $qb = $this->getDoctrine()->getManagerForClass($class)
```
I would create and set the QueryBulider in `deleteBatchAction` method itself. And so just send array $ids to `deleteBatchAction`.
IMHO, custom batch actions will often require more than a single DQL update/delete.

yceruto replied:
Still you can, just don't use the first argument but the second one passed to the custom method. It should cover both cases, simple and complex batch queries.

yceruto replied:
Done!

alterphp commented on src/Controller/AdminControllerTrait.php:
```diff
@@ -387,6 +387,36 @@ protected function searchAction()
+
+    protected function deleteBatchAction(QueryBuilder $qb): void
+    {
+        $qb->delete()->getQuery()->execute();
```
As said above, I would do all ORM action in this method.

I think throwing `PRE_BATCH_DELETE,` `POST_BATCH_DELETE` events will be useful too, with array `$ids` in the event too.

yceruto replied:
> I think throwing PRE_BATCH_DELETE, POST_BATCH_DELETE events will be useful too, with array $ids as event argument.

Not sure about this, this is just one line to override.

yceruto commented on src/Configuration/NormalizerConfigPass.php:
```diff
@@ -335,6 +336,28 @@ private function normalizeActionConfig(array $backendConfig)
 
+    private function normalizeBatchActionConfig(array $backendConfig)
+    {
+        if (!isset($backendConfig['list']['actions'])) {
```
`batch_actions`

yceruto commented on tests/Controller/DefaultBackendTest.php:
```diff
@@ -128,7 +128,7 @@ public function testListViewSearchAction()
 
         $this->assertSame('Search', \trim($crawler->filter('.form-action-search input[type="search"]')->attr('placeholder')));
-        $this->assertContains('form-action-search', $crawler->filter('.global-actions > div')->first()->attr('class'));
+        $this->assertContains('form-action-search', $crawler->filter('.global-actions > .global-actions > div')->first()->attr('class'));
```
This is a workaround to make test pass, later change the CSS class instead.

Pierstoval commented:
I really like the idea!

However, I'm not fan at all (or "not fan anymore", don't remember 🤣) of the "magic" method naming that resolves automatically to `{name}BatchAction()`

I prefer having a method such as `batchAction(string $actionName);` instead, where `$actionName` equals the name in the configuration.

This allows creating custom methods that are owned by the developer while still having the EasyAdmin method documented & auto-completed.

IMO we should remove the "magic" in EasyAdmin (dynamic methods, etc.), because it introduces lots of undocumented complexity. Overriding a method should be the main entry point to customizing, instead of "magic" ones.

yceruto commented:
Hi @Pierstoval, thanks for your input. I'm not fan of "magic" methods eirther, but I don't think it's something  we can solve soon. The current convention of the bundle is calling to `batchAction()` from `indexAction()` so there is no chance to have `batchAction(string $actionName)`.

However, it seems that ppl don't like "magic" so much and https://github.com/EasyCorp/EasyAdminBundle/pull/2578#discussion_r247691291 proposes that too.

@alterphp I've addressed your comment, thanks.

yceruto commented:
PR description updated with latest changes, it should feel less "magic" now.

alterphp commented on src/Resources/views/default/list.html.twig:
```diff
@@ -76,6 +83,25 @@
+    {% if _list_batch_actions is not empty %}
+        <div class="batch-actions" style="display: none">
+            <form method="post" action="{{ path('easyadmin', {action: 'batch', entity: _request_parameters.entity}) }}">
+                <input type="hidden" name="batch_action_ids">
```
Placing a Twig block in order to add some inputs (like a status selector near a "update status" button) would be useful, what do you think ?

yceruto replied:
In that case I'd prefer move the form definition to some controller/method and make there the extension point, thus you could add custom batch-fields/form-theme/custom-rendering, etc. WDYT?

yceruto replied:
Done!

alterphp commented on src/Resources/views/default/list.html.twig:
```diff
@@ -76,6 +83,25 @@
+    {% if _list_batch_actions is not empty %}
+        <div class="batch-actions" style="display: none">
+            <form method="post" action="{{ path('easyadmin', {action: 'batch', entity: _request_parameters.entity}) }}">
+                <input type="hidden" name="batch_action_ids">
```
Placing a Twig block in order to add some inputs (like a status selector near a "update status" button) would be useful, what do you think ?

alterphp commented:
> However, I'm not fan at all (or "not fan anymore", don't remember rofl) of the "magic" method naming that resolves automatically to `{name}BatchAction()`

I don't like it, but I think it's important in order to ease adoption of the bundle by less experimented developers and it increases the amount of installations. So finally, even if I don't use it at all, it could be useful for me (more people using the bundle implies bigger community, more contributors and so on).

alterphp commented:
Would be even better ! And pass $ids and form data to the (name)BatchAction.

Le mar. 15 janv. 2019 à 18:49, Yonel Ceruto <notifications@github.com> a
écrit :

> *@yceruto* commented on this pull request.
> ------------------------------
>
> In src/Resources/views/default/list.html.twig
> <https://github.com/EasyCorp/EasyAdminBundle/pull/2578#discussion_r247996489>
> :
>
> > @@ -76,6 +83,25 @@
>              </div>
>          {% endblock new_action %}
>      {% endif %}
> +    </div>
> +
> +    {% if _list_batch_actions is not empty %}
> +        <div class="batch-actions" style="display: none">
> +            <form method="post" action="{{ path('easyadmin', {action: 'batch', entity: _request_parameters.entity}) }}">
> +                <input type="hidden" name="batch_action_ids">
>
> In that case I'd prefer move the form definition to some controller/method
> and make there the extension point, thus you could add custom
> batch-fields/form-theme/custom-rendering, etc. WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/EasyCorp/EasyAdminBundle/pull/2578#discussion_r247996489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAtFzXGt2CSIoqa4KyBlN0iRcXej0Zb7ks5vDhTGgaJpZM4Z_lO5>
> .
>

yceruto commented on src/Resources/views/default/layout.html.twig:
```diff
@@ -111,30 +111,28 @@
 
                 {% block content %}
                     <div class="content">
-                        {% if block('content_header')|trim is not empty %}
```
This check makes it impossible to render a form within this block the second time it's rendered, because a form is rendered once.

alterphp replied:
I don't understand the test here, nor the next one on `global_actions`.

yceruto replied:
I just created a new PR to fix this one and other similar tests.
https://github.com/EasyCorp/EasyAdminBundle/pull/2585

yceruto commented on src/Resources/views/default/layout.html.twig:
```diff
@@ -111,30 +111,28 @@
-                            {% endblock content_header %}
-                        </section>
-                        {% endif %}
+                        {% block base_content_header %}
```
That's why I change it by a new block `base_content_header` wrapping the content, thus we've the same effect (overriding it and let empty) and better performance as all blocks are rendered once.

Same change for `{% if block('global_actions')|trim is not empty %}`, which was being rendered up to 4 times.

alterphp replied:
Ok, I think you're right, the test (is the block empty) indicates that it requires a wrapping block => `content_header_wrapper` insteadof of `base_content_header` ?

yceruto replied:
Better, thanks.

yceruto commented:
**Update:** 
 * Added new form type `EasyAdminBatchFormType` to define the batch form and batch fields/buttons. It'll allow you a custom batch form with inputs/dropdown/etc.. and custom rendering overriding `block easyadmin_batch_widget` if needed.
 * Added `{% block batch_actions %}` in `list.html.twig` template for easy customization.
 * Added method `createBatchForm()` in `AdminControllerTrait` for batch form creation and easy customization. The `batchAction()` method is a bit more readable now.

yceruto commented on src/Configuration/ActionConfigPass.php:
```diff
@@ -32,8 +32,11 @@ public function process(array $backendConfig)
         $backendConfig = $this->resolveActionInheritance($backendConfig);
+        $backendConfig = $this->resolveBatchActionInheritance($backendConfig);
         $backendConfig = $this->processActionsConfig($backendConfig);
+        $backendConfig = $this->processBatchActionsConfig($backendConfig);
```
(To facilitate the review) all new code related to this file is a copy-paste of `*ActionsConfig()` methods, removing the `foreach($this->views as $view)` statements and hardcoding the `list` view value.

yceruto commented on src/Resources/views/default/list.html.twig:
```diff
@@ -24,6 +24,7 @@
 {% endif %}
 
 {% set _request_parameters = _request_parameters|merge({ referer: path('easyadmin', _request_parameters)|url_encode }) %}
+{% set _has_batch_actions = _entity_config.list.batch_actions|length > 0 %}
```
FIXME: This actions aren't filtered by `disabled_actions`. 

Likely this static process (this one: https://github.com/EasyCorp/EasyAdminBundle/blob/524f9310814436a041419cc6f3b985e6814e402f/src/Twig/EasyAdminTwigExtension.php#L331-L350) should be done by `ActionConfigPass`.

yceruto replied:
Fixed. Reusing `batch_form.vars.batch_actions` var, which was already filtered in `buildView()`.

yceruto commented on src/Form/Type/EasyAdminBatchFormType.php:
```diff
@@ -0,0 +1,82 @@
+
+        $view->vars['batch_actions'] =  \array_filter($batchActions, function ($batchAction) use ($disabledActions) {
+            return !\in_array($batchAction['name'], $disabledActions, true);
+        });
```
So we could remove this repetitive task everywhere, @javiereguiluz what do you think?

yceruto replied:
Alternative solution applied.

yceruto commented on src/Resources/views/form/bootstrap_4.html.twig:
```diff
@@ -478,3 +478,16 @@
+{% block easyadmin_batch_widget %}
+    <button type="button" class="btn btn-link deselect-batch-button">Deselect</button>
+    {% for action in batch_actions %}
+        <button type="submit" class="btn {{ action.css_class|default('') }}" title="{{ action.title|default('') is empty ? '' : action.title|trans(_trans_parameters) }}" name="{{ form.name.vars.full_name }}" value="{{ action.name }}">
```
Or better, use default `|default('btn-secondary')` here.

alterphp commented on src/Resources/config/form.xml:
```diff
@@ -15,6 +15,11 @@
             <tag name="form.type" alias="easyadmin_autocomplete" />
         </service>
 
+        <service id="easyadmin.form.type.batch" class="EasyCorp\Bundle\EasyAdminBundle\Form\Type\EasyAdminBatchFormType">
```
I think we should now use FQCN for new service (appeared from 2.0 version) ? @javiereguiluz any opinion about this ?

yceruto replied:
If so, I think we should do these changes together.

yceruto replied:
I'd prefer consistency here :)

I'm also not sure if it's a good practice for third-party bundles, unless the service is clearly useful for the end developer at project level, plus if you've `autowire` option activated in your app.

alterphp replied:
Right

yceruto commented:
Hey! demo project ready! https://github.com/yceruto/ea2_batch_actions check it out.

See diff https://github.com/yceruto/ea2_batch_actions/commit/2f228668cf99df612e6738d753a2aa785e065097 to know what changes you'll need to create a batch action.

alterphp commented:
To me it's ready to merge ! @javiereguiluz WDYT ?

yceruto commented:
I just pushed a new branch: https://github.com/yceruto/ea2_batch_actions/tree/custom_batch_form with custom batch form. See diff https://github.com/yceruto/ea2_batch_actions/commit/f720a88a7523f4352f74be8e4ef2209f02e74fdd

![eas2-status-batch-actions](https://user-images.githubusercontent.com/2028198/51392117-076d5100-1b02-11e9-94b7-1dea67c223c8.gif)

@alterphp It was your idea, tell me how it looks for you?

And this feature clears the road for this card too: https://github.com/EasyCorp/EasyAdminBundle/projects/1#card-1953061 (Export to CSV, Excel, JSON, PDF).

yceruto commented:
@alterphp not yet, see TODO list :) also we need the @javiereguiluz's approval to continue.

alterphp commented:
CSS is broken on my side, did not manage to get it right even after compiling EAB vendor styles. aNything I'm missing ?

yceruto commented:
@alterphp It should be fixed now.

yceruto commented:
(Rebased) This is ready for review.

javiereguiluz commented:
@yceruto thanks a lot for contributing this feature. I love it. I'm going to review it soon and merge it. In your description you say that "docs" and "tests" are left to do.

I propose you something: let's finish this feature making sure all existing tests pass and merge it. Later, I can work on both tests and docs. That way I'll learn how this works and could propose changes/tweaks if needed. Thanks.

javiereguiluz commented:
Thanks a lot Yonel. This is now merged and I'll play with it to create the docs and tests. Thanks.

yceruto commented:
Very excited about this, thank you!!

TomasVotruba commented:
Amazing work :clap:

TomasVotruba commented:
I could not find it, but when batch action doesn't return `Response`, how can I return `FileResponse`?

I'm downloading multiple PDF files wrapped to ZIP with batch

yceruto commented:
Hi @TomasVotruba, that feature is not currently supported, but it seems to me a valid use case.

https://github.com/EasyCorp/EasyAdminBundle/blob/6ac97da2a3bff834a195653882c9c4a778f9d56a/src/Controller/AdminControllerTrait.php#L401-L414

Just would need to check whether the called method return a `Response` instance or not, if so return it, otherwise `$this->redirectToReferrer();`

Do you like to make the changes?

TomasVotruba commented:
Hi @yceruto ,

thanks for quick reply!

I tried to do them locally, but missing covariance doesn't allow me to change `RedirectResponse`

TomasVotruba commented:
Done: https://github.com/EasyCorp/EasyAdminBundle/pull/2732
